# Quantum Mechanics: The Wave Function

## Separable Variable Wave Function

### First Wave Function

In this section we are going to derive and plot a simple separable wave function that solves the time-independent Schrödinger equation ($E\Psi = \hat{H}\Psi$). These solutions were found by using the separation of variables technique on the Schrödinger equation. The solutions are all in the following form,

$$\Psi(x,t)=\psi(x)e^{-\frac{iE}{\hbar}t},$$ 

with \psi being the time independent wave function. For this first example we have chosen it to simply be 

$$\psi(x)=Ae^{-x^2}$$

with A being the normalization factor. The following is done to normalize the probability density function:

$$ \int_{-\infty}^\infty \left|\Psi(x,t)\right|^2\mathrm{d}x\mathrm{d}t = 1 $$

$$ \int_{-\infty}^\infty \left|\psi(x)e^{-\frac{iE}{\hbar}t}\right|^2\mathrm{d}x\mathrm{d}t = 1 $$

$$ \int_{-\infty}^\infty \left|Ae^{-x^2}e^{-\frac{iE}{\hbar}t}\right|^2\mathrm{d}x\mathrm{d}t = 1 $$

$$ \int_{-\infty}^\infty \left|Ae^{-x^2}\right|^2\mathrm{d}x\int_{-\infty}^\infty\left|e^{-\frac{iE}{\hbar}t}\right|^2\mathrm{d}t = 1 $$

here $\int_{-\infty}^\infty\left|e^{-\frac{iE}{\hbar}t}\right|^2\mathrm{d}t = 1$ because it's a complex square and so cancels. So we end up with,

$$ \int_{-\infty}^\infty \left|Ae^{-x^2}\right|^2\mathrm{d}x = 1 $$

$$ \left|A\right|^2 \int_{-\infty}^\infty \left|e^{-x^2}\right|^2\mathrm{d}x = 1 $$

$$ \left|A\right|^2 \int_{-\infty}^\infty e^{-2x^2}\mathrm{d}x = 1$$

From previous solutions, we know that $\int_{-\infty}^\infty e^{-2x^2}\mathrm{d}x = \sqrt{\frac{\pi}{2}}$ and so,

$$ \left|A\right|^2 \sqrt{\frac{\pi}{2}} = 1 $$

$$ A = \left(\frac{2}{\pi}\right)^{\frac{1}{4}} $$

So we now have the full wave function, 

$$\Psi(x,t)=\left(\frac{2}{\pi}\right)^{\frac{1}{4}}e^{-x^2}e^{-\frac{iE}{\hbar}t},$$ 

which can also be broken down into the real and imaginary parts, that are more useful for the next part:

$$\Psi(x,t)=\left(\frac{2}{\pi}\right)^{\frac{1}{4}}e^{-x^2}\left(\cos\left(\frac{Et}{\hbar}\right)-i\sin\left(\frac{Et}{\hbar}\right)\right)$$

In [0]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

from matplotlib import animation, rc
from IPython.display import HTML

In [0]:
def init1():
    line1_1.set_data([], [])
    line1_2.set_data([], [])
    line1_3.set_data([], [])
    return line1_1,line1_2,line1_3

def animate1(i):
    #super important to clear the legends every image!
    ax1[1].collections.clear()
    
    x = np.linspace(-5, 5, 1000)
    psi1_Re = ((2/np.pi)**(1/4))*np.exp(-x**2)*np.cos((2*np.pi/100)*i)
    psi1_Im = -((2/np.pi)**(1/4))*np.exp(-x**2)*np.sin((2*np.pi/100)*i)
    p_dens1 = np.abs(psi1_Re+1j*psi1_Im)**2
    
    ax1[1].fill_between(x,p_dens1,facecolor='lightcoral', alpha=0.3,label=r"$\int_{-\infty}^\infty \left\|\Psi(x,t)\right|^2\mathrm{d}x$")
    
    ax1[0].text(2.4, 5.5, r"$E\Psi = \hat{H}\Psi$",horizontalalignment='center', fontsize=13)
    ax1[1].text(2.4, 5.5, r"$E\Psi = \hat{H}\Psi$",horizontalalignment='center', fontsize=13)
    
    ax1[0].legend()
    ax1[1].legend()
    ax1[0].set_yticklabels([])
    ax1[0].set_xticklabels([])
    ax1[1].set_yticklabels([])
    ax1[1].set_xticklabels([])

    line1_1.set_data(x, psi1_Re)
    line1_2.set_data(x, psi1_Im)
    line1_3.set_data(x, p_dens1)
    
    return line1_1,line1_2,line1_3

In [0]:
%%capture
fig1, ax1 = plt.subplots(1,2, figsize=(14, 6),constrained_layout=True)

ax1[0].set_ylabel(r"$\Psi$")
ax1[1].set_ylabel(r"$\left\|\Psi\right|^2$")

fig1.suptitle("Wave function :" + r"$\Psi(x,t)=\sqrt[4]{\frac{2}{\pi}}e^{-x^2}e^{\frac{-iE}{\hbar}t}$", fontsize=16)
ax1[0].set_title("Evolution of the wave function " + r"$\Psi(x,t)$")
ax1[1].set_title("Probability density function for position of particle")

ax1[0].set_xlim(( -4, 4))
ax1[0].set_ylim((-1.1, 1.7))

ax1[1].set_xlim(( -3, 3))
ax1[1].set_ylim((-.5, 1.5))

line1_1, = ax1[0].plot([], [], lw=2, label=r"$\Re(\Psi(x,t))=\sqrt[4]{\frac{2}{\pi}}e^{-x^2}\cos(\frac{Et}{\hbar})$", color="blue")
line1_2, = ax1[0].plot([], [], lw=2, label=r"$\Im(\Psi(x,t))=-\sqrt[4]{\frac{2}{\pi}}e^{-x^2}i\sin(\frac{Et}{\hbar})$", color="orange")
line1_3, = ax1[1].plot([], [], lw=2, label=r"$\left\|\Psi(x,t)\right|^2$", color="red")

plt.subplots_adjust(left=0.1, right=0.9, top=0.85, bottom=0.1)

In [0]:
anim = animation.FuncAnimation(fig1, animate1, init_func=init1,frames=100, interval=20, blit=True)

HTML(anim.to_html5_video())

We can see that the probability distribution is constant: this is how a stationnary state is described! Every solution found by separation of variables is like this: probability density does not depend on time, nor does any expectation value: they are constant in time. Let's see with the next wave function... 

### Second Wave Function

Here we just multiply $\psi(x)=Ae^{-x^2}$ with $x$, to get 

$$\psi(x)=Ae^{-x^2},$$

our second time-independant wave function. We just repeate the same steps as above, and discover that this time, 

$$ A = \left(\frac{32}{\pi}\right)^{\frac{1}{4}} $$

As such, we get our new wave function, 

$$\Psi(x,t)=\left(\frac{32}{\pi}\right)^{\frac{1}{4}}xe^{-x^2}e^{-\frac{iE}{\hbar}t},$$ 

$$\Psi(x,t)=\left(\frac{32}{\pi}\right)^{\frac{1}{4}}xe^{-x^2}\left(\cos\left(\frac{Et}{\hbar}\right)-i\sin\left(\frac{Et}{\hbar}\right)\right)$$

And so animated, it gives something like: 

In [0]:
def init2():
    line2_1.set_data([], [])
    line2_2.set_data([], [])
    line2_3.set_data([], [])
    return line2_1,line2_2,line2_3

def animate2(i):
    ax2[1].collections.clear()
    
    x = np.linspace(-5, 5, 1000)
    psi2_Re = ((32/np.pi)**(1/4))*x*np.exp(-x**2)*np.cos((2*np.pi/100)*i)
    psi2_Im = -((32/np.pi)**(1/4))*x*np.exp(-x**2)*np.sin((2*np.pi/100)*i)
    p_dens2 = np.abs(psi2_Re+1j*psi2_Im)**2
    
    ax2[1].fill_between(x,p_dens2,facecolor='lightcoral', alpha=0.3,label=r"$\int_{-\infty}^\infty \left\|\Psi(x,t)\right|^2\mathrm{d}x$")
    
    ax2[0].text(2.4, 5.5, r"$E\Psi = \hat{H}\Psi$",horizontalalignment='center', fontsize=13)
    ax2[1].text(2.4, 5.5, r"$E\Psi = \hat{H}\Psi$",horizontalalignment='center', fontsize=13)
    
    ax2[0].legend()
    ax2[1].legend()
    ax2[0].set_yticklabels([])
    ax2[0].set_xticklabels([])
    ax2[1].set_yticklabels([])
    ax2[1].set_xticklabels([])

    line2_1.set_data(x, psi2_Re)
    line2_2.set_data(x, psi2_Im)
    line2_3.set_data(x, p_dens2)
    
    return line2_1,line2_2,line2_3

In [0]:
%%capture
fig2, ax2 = plt.subplots(1,2, figsize=(12, 6),constrained_layout=True)

ax2[0].set_ylabel(r"$\Psi$")
ax2[1].set_ylabel(r"$\left\|\Psi\right|^2$")

fig2.suptitle("Wave function :" + r"$\Psi(x,t)=\sqrt[4]{\frac{32}{\pi}}xe^{-x^2}e^{\frac{-iE}{\hbar}t}$", fontsize=16)
ax2[0].set_title("Evolution of the wave function " + r"$\Psi(x,t)$")
ax2[1].set_title("Probability density function for position of particle")

ax2[0].set_xlim(( -4, 4))
ax2[0].set_ylim((-1.1, 1.7))

ax2[1].set_xlim(( -3, 3))
ax2[1].set_ylim((-.5, 1.5))

line2_1, = ax2[0].plot([], [], lw=2, label=r"$\Re(\Psi(x,t))=\sqrt[4]{\frac{32}{\pi}}xe^{-x^2}\cos(\frac{Et}{\hbar})$", color="blue")
line2_2, = ax2[0].plot([], [], lw=2, label=r"$\Im(\Psi(x,t))=-\sqrt[4]{\frac{32}{\pi}}xe^{-x^2}i\sin(\frac{Et}{\hbar})$", color="orange")
line2_3, = ax2[1].plot([], [], lw=2, label=r"$\left\|\Psi(x,t)\right|^2$", color="red")

plt.subplots_adjust(left=0.1, right=0.9, top=0.85, bottom=0.1)

In [0]:
anim2 = animation.FuncAnimation(fig2, animate2, init_func=init2,frames=100, interval=20, blit=True)

HTML(anim2.to_html5_video())

And once again, we see that with a separated variable solution, we end up with constant prob. distibution and all that goes with it.

$$ \Psi_0(x) = e^{-\frac{\omega}{2\hbar}x^2} $$
$$ \Psi_1(x) = 2i\omega xe^{-\frac{\omega}{2\hbar}x^2} $$
$$ \Psi_2(x) = (-\hbar+2\omega x^2)e^{-\frac{\omega}{2\hbar}x^2} $$


In [2]:
import sympy as sp
from sympy import oo, I

hbar = 1
omega = 1

x = sp.Symbol('x')
def normalize(psi, omega):
  if psi == 0:
    return 1/sp.sqrt((sp.integrate(sp.Abs(sp.exp(-(omega/(2*hbar))*(x**2)))**2,(x,-oo,oo))))
  
  elif psi == 1:
    return sp.S(sp.sqrt(sp.integrate(sp.Abs(2*I*omega*x*sp.exp(-(omega/(2*hbar)*(x**2))))**2,(x,-oo,oo)))**(-1),3)

  elif psi == 2:
    return sp.sqrt(sp.integrate(sp.Abs((-hbar+2*omega*x**2)*sp.exp(-(omega/(2*hbar)*(x**2))))**2,(x,-oo,oo)))**(-1)

N0 = normalize(0,omega)
N1 = normalize(1,omega)
N2 = normalize(2,omega)
#WE NEED TO CONVERT TO NORMAL FLOATS, OTHERWISE MESSES UP STUFF - SORRY THIS IS VERY UGLY BUT I COULD NOT STAND THESE DIFFERENT OBJECT TYPES NOT WORKING TOGETHER...
N_ = [round(float(sp.N(N0)),4),   round(float(sp.N(N1)),4),   round(float(sp.N(N2)),4)]

#LATEX TEXT FOR THE FUNCTION
la_N = ["$"+str(N_[0])+r"e^{-\frac{\omega}{2\hbar}x^2}$","$"+str(N_[1])+r"2i\omega xe^{-\frac{\omega}{2\hbar}x^2}$","$"+str(N_[2])+r"(-\hbar+2\omega x^2)e^{-\frac{\omega}{2\hbar}x^2}$"]

#THE DIFFERENT WAYS WE CAN OUTPUT THE VALUES
print(N0)
print(la_N[0])
print(N_[0])

1.0/pi**(1/4)
$0.7511e^{-\frac{\omega}{2\hbar}x^2}$
0.7511


In [0]:
def init3():
  for j in range(3):
    line3_1[j].set_data([], [])
    line3_2[j].set_data([], [])
    line3_3[j].set_data([], [])
    return line3_1[0],line3_1[1],line3_1[2],line3_2[0],line3_2[1],line3_2[2],line3_3[0],line3_3[1],line3_3[2]

def animate3(i):
  xarr = np.linspace(-5, 5, 1000)
  psi3_Re = [N_[0]*np.exp(-(omega/(2*hbar))*xarr**(2))*np.cos((2*np.pi/100)*i),   N_[1]*2*omega*xarr*np.exp(-(omega/(2*hbar))*xarr**(2))*np.cos((2*np.pi/100)*i),   N_[2]*(-hbar+2*omega*(xarr**2))*np.exp(-(omega/(2*hbar))*xarr**(2))*np.cos((2*np.pi/100)*i)]
  psi3_Im = [-N_[0]*np.exp(-(omega/(2*hbar))*xarr**(2))*np.sin((2*np.pi/100)*i),   -N_[1]*2*omega*xarr*np.exp(-(omega/(2*hbar))*xarr**(2))*np.sin((2*np.pi/100)*i),    -N_[2]*(-hbar+2*omega*(xarr**2))*np.exp(-(omega/(2*hbar))*xarr**(2))*np.sin((2*np.pi/100)*i)]
  p_dens3 = [np.abs(psi3_Re[0]+1j*psi3_Im[0])**2,np.abs(psi3_Re[1]+1j*psi3_Im[1])**2,np.abs(psi3_Re[2]+1j*psi3_Im[2])**2]
    
  for j in range(3):
    ax3[j][1].collections.clear()
    
    ax3[j][1].fill_between(xarr,p_dens3[j],facecolor='lightcoral', alpha=0.3,label=r"$\int_{-\infty}^\infty \left\|\Psi(x,t)\right|^2\mathrm{d}x$")
    
    ax3[j][0].legend(loc='upper right',fontsize=8)
    ax3[j][1].legend(loc='upper right',fontsize=8)
    ax3[j][0].set_yticklabels([])
    ax3[j][0].set_xticklabels([])
    ax3[j][1].set_yticklabels([])
    ax3[j][1].set_xticklabels([])

    line3_1[j].set_data(xarr, psi3_Re[j])
    line3_2[j].set_data(xarr, psi3_Im[j])
    line3_3[j].set_data(xarr, p_dens3[j])
    
  return line3_1[0],line3_1[1],line3_1[2],line3_2[0],line3_2[1],line3_2[2],line3_3[0],line3_3[1],line3_3[2]

In [0]:
%%capture

fig3, ax3 = plt.subplots(3,2, figsize=(12, 2*6),constrained_layout=True)
fig3.suptitle("Wave functions of a Quantum Harmonic Oscillator :\n" + r"$\Psi_N(x) = \psi_N(x)e^{\frac{-iE}{\hbar}t}$", fontsize=16)

line3_1=[None,None,None]
line3_2=[None,None,None]
line3_3=[None,None,None]

for k in range(3):
  ax3[k][0].set_title("Evolution of the wave function\n" + r"$\Psi_{}(x,t) = $".format(k) + r"{}".format(la_N[k]) + r"$e^{-\frac{\omega}{2\hbar}x^2}$")
  ax3[k][1].set_title("Probability density function for position of particle")

  ax3[k][0].set_ylabel(r"$\Psi$")
  ax3[k][1].set_ylabel(r"$\left\|\Psi\right|^2$")

  ax3[k][0].set_xlim(( -4, 4))
  ax3[k][0].set_ylim((-1.1, 1.7))

  ax3[k][1].set_xlim(( -4, 4))
  ax3[k][1].set_ylim((-.5, 1.5))

  line3_1[k], = ax3[k][0].plot([], [], lw=2, label=r"$\Re(\Psi(x,t))=$"+r"{}".format(la_N[k])+r"$\cos(\frac{Et}{\hbar})$", color="blue")
  line3_2[k], = ax3[k][0].plot([], [], lw=2, label=r"$\Im(\Psi(x,t))=$"+r"{}".format(la_N[k])+r"$i\sin(\frac{Et}{\hbar})$", color="orange")
  line3_3[k], = ax3[k][1].plot([], [], lw=2, label=r"$\left\|\Psi(x,t)\right|^2$", color="red")

plt.subplots_adjust(left=0.1, right=0.9, top=0.88, bottom=0.1, hspace=0.5)

In [6]:
anim3 = animation.FuncAnimation(fig3, animate3, init_func=init3,frames=500, interval=10, blit=True)

#anim3.save('QuantHarmOscillator.mp4',dpi=600)
HTML(anim3.to_html5_video())

The symbolic system on Sympy did not manage to write down nicely the normalization constants values, so we will do it here:

\begin{align*}
&\Psi_0(x) = e^{-\frac{\omega}{2\hbar}x^2}\\
&\Psi_1(x) = 2i\omega xe^{-\frac{\omega}{2\hbar}x^2}\\
&\Psi_2(x) = (-\hbar+2\omega x^2)e^{-\frac{\omega}{2\hbar}x^2}\\
\end{align*}

So in order, the solving of the normalization constants go as follows, 

\begin{align*}
\int_{-\infty}^\infty\left|\Psi_0(x,t)\right|^2
&= \int_{-\infty}^\infty\left| A  e^{-\frac{\omega}{2\hbar}x^2}e^{\frac{-iE}{\hbar}t} \right|^2 dx\\
1 &= |A|^2\int_{-\infty}^\infty\left| e^{-\frac{\omega}{2\hbar}x^2} \right|^2 dx\\
1 &= |A|^2 \sqrt\pi \\
A &= \frac{1}{\sqrt[4]\pi}\\
A &\approx 0.7511\\
\end{align*}
<br>
\begin{align*}
\int_{-\infty}^\infty\left|\Psi_1(x,t)\right|^2
&= \int_{-\infty}^\infty\left| A\cdot2i\omega xe^{-\frac{\omega}{2\hbar}x^2}e^{\frac{-iE}{\hbar}t} \right|^2 dx\\
1 &= |A|^2\int_{-\infty}^\infty\left| 2i\omega xe^{-\frac{\omega}{2\hbar}x^2} \right|^2 dx\\
1 &= |A|^2 \cdot 2\sqrt\pi \\
A &= \frac{1}{\sqrt2\sqrt[4]\pi}\\
A &\approx 0.53112\\
\end{align*}
<br>
\begin{align*}
\int_{-\infty}^\infty\left|\Psi_2(x,t)\right|^2
&= \int_{-\infty}^\infty\left| A\cdot2i\omega xe^{-\frac{\omega}{2\hbar}x^2}e^{\frac{-iE}{\hbar}t} \right|^2 dx\\
1 &= |A|^2\int_{-\infty}^\infty\left| 2i\omega xe^{-\frac{\omega}{2\hbar}x^2} \right|^2 dx\\
1 &= |A|^2 \cdot 2\sqrt\pi \\
A &= \frac{1}{\sqrt2\sqrt[4]\pi}\\
A &\approx 0.53112\\
\end{align*}

Also to note, $\omega$ and $\hbar$ are variables in the code. At the moment they are both equal to 1, as in this context it is not really necessary to have their real value. 